docs/_posts/2022-06-24-types-as-comments-proposal.md

```diff
@@ -0,0 +1,170 @@
+
+```bash
+$ node -e 'console.log(3 + [])'
+3
+$ node -e 'console.log({}.foo)' 
+undefined
+```
+
+> :information_source: Have a look at [Programming TypeScript][oreilly-programming-typescript] (p. 3) for more examples.
+> 
+
+Obviously, even though the `+` operator may be overloaded to add strings and arrays, it doesn't make a lot of sense to do that in the first place. Neither does accessing a custom property `foo` on a, very obviously, empty object. The developers of JavaScript, in a very questionable attempt to make developer lives easier in the short run, opted to take care of these kinds of problems by
+
+1. in the first case applying [type coercion](https://developer.mozilla.org/en-US/docs/Glossary/Type_coercion), and 
+2. in the second case by defaulting to the special value `undefined` that models undefined values.
+
+If you have ever ripped your hairs trying to find the cause of the infamous "TypeError: Cannot read property 'bar' of undefined" runtime error, e.g.
+```bash
+$ node -e 'console.log({}.foo.bar)'
+[eval]:1
+console.log({}.foo.bar)
+                   ^
+
+TypeError: Cannot read property 'bar' of undefined
+    at [eval]:1:20
+    at Script.runInThisContext (vm.js:134:12)
+    at Object.runInThisContext (vm.js:310:38)
+    at internal/process/execution.js:81:19
+    at [eval]-wrapper:6:22
+    at evalScript (internal/process/execution.js:80:60)
+    at internal/main/eval_string.js:27:3
+```
+you probably understand why, in the long run, you want to catch these kinds of problems **before** running your code. Preferably even, you want to know this immediately when you type the code into your editor of choice.
+
+### TypeScript to the rescue
+
+Part of the appeal of TypeScript is that it adds this additional layer of static—i.e. before runtime—safety to your JavaScript code.
+
+> "The main benefit of TypeScript is that it can highlight unexpected behavior in your code, lowering the chance of bugs."
+> 
+> — [TypeScript in 5 Minutes](https://www.typescriptlang.org/docs/handbook/typescript-in-5-minutes.html)
+
+Let's now, for example, directly run the above code examples with TypeScript (using `ts-node`).
+
+```bash
+$ ts-node -e 'console.log(3 + [])'
+[eval].ts:1:13 - error TS2365: Operator '+' cannot be applied to types 'number' and 'never[]'.
+
+1 console.log(3 + [])
+              ~~~~~~
+$ ts-node -e 'console.log({}.foo.bar)'
+[eval].ts:1:16 - error TS2339: Property 'foo' does not exist on type '{}'.
+
+1 console.log({}.foo.bar)
+                 ~~~
+```
+
+As we can see, this type of error can easily be detected before running the code by applying type checking. In the first case, TypeScript tells us that we should reconsider adding a number and an array. In the second case, we are informed that empty objects do not magically come with an arbitrary property called `foo`. Both are very reasonable properties of safe JavaScript code and it's extremely helpful that TypeScript checks them for us before we run the code.
```

Comment:
I'd change the wording

```suggestion
As we can see, this type of error can easily be detected before running the code by applying type checking. In the first case, TypeScript tells us that we can not add or concatenate a number and an array. In the second case, we are informed that empty objects do not magically come with an arbitrary property called `foo`. Both are very reasonable properties of safe JavaScript code and it's extremely helpful that TypeScript checks them for us before we run the code.
```

Reply:
We actually **can** do that in JavaScript tho 😅 it makes very little sense but we can do that.